### PR TITLE
imp: add default methods to BeanDefinition to ease checking whether a method is annotated  

### PR DESCRIPTION
**Pull request by sdelamo**
see: https://github.com/micronaut-projects/micronaut-security/pull/1526

### CODE REVIEW

**Review by graemerocher — CHANGES_REQUESTED** [review]
I don't agree with exposing yet another public set of static utility methods which will confuse users and the API. Seems like these methods could added as default methods to `BeanRegistration` / `BeanDefinition` interface if you want to simplify the logic bit

**Comment by sdelamo** [comment]
> I don't agree with exposing yet another public set of static utility methods which will confuse users and the API. Seems like these methods could added as default methods to `BeanRegistration` / `BeanDefinition` interface if you want to simplify the logic bit

I have removed the `Utils` classes.

File: inject/src/main/java/io/micronaut/inject/BeanDefinition.java [hunk 1]
```diff
@@ -253,6 +254,30 @@ default <R> Optional<ExecutableMethod<T, R>> findMethod(String name, Class<?>...
...
 
+    /**
+     * Whether an {@link ExecutableMethod} exists which is annotated with the supplied annotation
+     *
+     * @param methodName          The method name
+     * @param argumentTypes The argument types
+     * @return Whether an {@link ExecutableMethod} exists which is annotated with the supplied annotation
+     * @since 4.3.0
+     */
+    default boolean existsMethodWithAnnotation(Class<? extends Annotation> annotationClass, String methodName, Class<?>... argumentTypes) {
```
Comment (graemerocher) [inline inject/src/main/java/io/micronaut/inject/BeanDefinition.java]: rename to `hasAnnotatedMethod`

File: inject/src/main/java/io/micronaut/inject/BeanDefinition.java [hunk 2]
```diff
@@ -253,6 +254,30 @@ default <R> Optional<ExecutableMethod<T, R>> findMethod(String name, Class<?>...
...
+
+    /**
+     * Whether an {@link ExecutableMethod} exists which is annotated with {@link Blocking}.
+     *
+     * @param methodName          The method name
+     * @param argumentTypes The argument types
+     * @return Whether an {@link ExecutableMethod} exists which is annotated with with {@link Blocking}.
+     * @since 4.3.0
+     */
+    default boolean existsMethodAnnotatedWithBlocking(String methodName, Class<?>... argumentTypes) {
```
Comment (graemerocher) [inline inject/src/main/java/io/micronaut/inject/BeanDefinition.java]: I don't think this method adds any value, and we shouldn't have methods for every annotation type here. Just remove it and you can pass `Blocking.class` when calling the above method

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: inject/src/main/java/io/micronaut/inject/BeanDefinition.java (hunk 1 above)
Comment (graemerocher) [reply inject/src/main/java/io/micronaut/inject/BeanDefinition.java]: Also add nullability annotations to the arguments

**Comment by dstepanov** [comment]
I'm not sure if we should push forward the idea of annotating blocking services and let us do something about it, if we use virtual threads blocking is fine. It's better to have something like `@ExecuteOn`

File: inject/src/main/java/io/micronaut/inject/BeanDefinition.java (hunk 2 above)
Comment (sdelamo) [reply inject/src/main/java/io/micronaut/inject/BeanDefinition.java]: resolved via https://github.com/micronaut-projects/micronaut-core/pull/10261/commits/b02ad851014ab4a1a9d15cabb3acba13c3d1f114

**Review by graemerocher — APPROVED** [review]

**Comment by dstepanov** [comment]
Do we really want to add `findBeanDefinition(@NonNull T bean)`? That requires iterating all the beans

**Comment by sdelamo** [comment]
> Do we really want to add `findBeanDefinition(@NonNull T bean)`? That requires iterating all the beans

Yes, I think a method to get a bean definition given an instance it is necessary. Is there are more optimal way to do this?

Based on your comment I assume the issue is not this PR but that `BeanDefinitionRegistry::findBeanRegistration` implementation is not optimised for lookups and does iterate all the beans.

**Comment by dstepanov** [comment]
The best way would be to get a bean definition/s and after that get a bean instance.

**Comment by sdelamo** [comment]
> The best way would be to get a bean definition/s and after that get a bean instance.

but you may have a bean instance already.

**Comment by dstepanov** [comment]
Getting a bean definition from an instance should be avoided, maybe refactored the usage of it.

**Comment by sdelamo** [comment]
> Getting a bean definition from an instance should be avoided, maybe refactored the usage of it.

This PR is the usecase: https://github.com/micronaut-projects/micronaut-security/pull/1526

It seems you think we should not have the method `BeanDefinitionRegistry::findBeanRegistration(Object bean)`. Please note this PR does not add this method. It was already there.

**Comment by sdelamo** [comment]
@dstepanov can you approve this PR o should I close it?

**Review by dstepanov — APPROVED** [review]

**Pull request merged**